framework/source/class/qx/core/Id.js

```diff
@@ -0,0 +1,166 @@
+/* ************************************************************************
+
+   qooxdoo - the new era of web development
+
+   http://qooxdoo.org
+
+   Copyright:
+     2017 Zenesis Limited, http://www.zenesis.com
+
+   License:
+     MIT: https://opensource.org/licenses/MIT
+     See the LICENSE file in the project's top-level directory for details.
+
+   Authors:
+     * John Spackman (john.spackman@zenesis.com, @johnspackman)
+
+************************************************************************ */
+
+/**
+ * Provides a registry of top level objects
+ */
+qx.Class.define("qx.core.Id", {
+  extend: qx.core.Object,
+  include: [ qx.core.MObjectId ],
+  type: "singleton",
+  
+  members: {
+    
+    __registeredObjects: null,
+    __registeredIdHashes: null,
+
+    /*
+     * @Override
+     */
+    _createObjectImpl: function(id) {
+      if (this.__registeredObjects) {
+        var obj = this.__registeredObjects[id];
+        if (obj !== undefined) {
+          return obj;
+        }
+      }
+      
+      switch(id) {
+        case "application":
+          return qx.core.Init.getApplication() || undefined;
+      }
+      
+      return; // undefined
+    },
+    
+    /**
+     * Returns an object path which can be used to locate an object anywhere in the application
+     * with a call to `qx.core.Id.getObject()`.
+     * 
+     * This will return null if it is not possible to calculate a path because one of the
+     * ancestors has a null `objectId`.
+     * 
+     * This will also return null if the top-most ancestor is not one of the globals registered
+     * with `registerObject` or a known global (such as the application); however, by passing
+     * `false` as the `mustBeRegistered` parameter, this restriction is removed and a path will
+     * still be returned (it's just that in this case, you will have to know the the root object
+     * to query and `qx.core.Id.getObject` will not work on that path)  
+     * 
+     * @param obj {qx.core.Object} the object
+     * @param mustBeRegistered {Boolean} whether the
+     * @return {String} full path to the object
+     */
+    getAbsoluteIdOf: function(obj, mustBeRegistered) {
+      var segs = [];
+      var application = qx.core.Init.getApplication();
+      while (obj) {
+        var id = obj.getObjectId();
+        if (!id) {
+          this.error("Cannot determine an absolute Object ID because one of the ancestor ObjectID's is null");
```

Comment:
Would it be useful here, in the error message, to state which child's parent had a null object id?